File: src/Python/GeometricObjects/OrientedCylinder.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import vtk

'''
There are two alternative ways to apply the transform.
 1) Use vtkTransformPolyDataFilter to create a new transformed polydata.
    This method is useful if the transformed polydata is needed
      later in the pipeline
    To do this, set USER_MATRIX = True
 2) Apply the transform directly to the actor using vtkProp3D's SetUserMatrix.
    No new data is produced.
    To do this, set USER_MATRIX = False
'''
USER_MATRIX = True


def main():
    colors = vtk.vtkNamedColors()

    # Set the background color.
    colors.SetColor("BkgColor", [26, 51, 77, 255])

    # Create a cylinder.
    # Cylinder height vector is (0,1,0).
    # Cylinder center is in the middle of the cylinder
    cylinderSource = vtk.vtkCylinderSource()
    cylinderSource.SetResolution(15)

    # Generate a random start and end point
    startPoint = [0] * 3
    endPoint = [0] * 3
    rng = vtk.vtkMinimalStandardRandomSequence()
    rng.SetSeed(8775070)  # For testing.
    for i in range(0, 3):
        rng.Next()
        startPoint[i] = rng.GetRangeValue(-10, 10)
        rng.Next()
        endPoint[i] = rng.GetRangeValue(-10, 10)

    # Compute a basis
    normalizedX = [0] * 3
    normalizedY = [0] * 3
    normalizedZ = [0] * 3

    # The X axis is a vector from start to end
    vtk.vtkMath.Subtract(endPoint, startPoint, normalizedX)
    length = vtk.vtkMath.Norm(normalizedX)
    vtk.vtkMath.Normalize(normalizedX)

    # The Z axis is an arbitrary vector cross X
    arbitrary = [0] * 3
    for i in range(0, 3):
        rng.Next()
        arbitrary[i] = rng.GetRangeValue(-10, 10)
    vtk.vtkMath.Cross(normalizedX, arbitrary, normalizedZ)
    vtk.vtkMath.Normalize(normalizedZ)

    # The Y axis is Z cross X
    vtk.vtkMath.Cross(normalizedZ, normalizedX, normalizedY)
    matrix = vtk.vtkMatrix4x4()

    # Create the direction cosine matrix
    matrix.Identity()
    for i in range(0, 3):
        matrix.SetElement(i, 0, normalizedX[i])
        matrix.SetElement(i, 1, normalizedY[i])
        matrix.SetElement(i, 2, normalizedZ[i])

    # Apply the transforms
    transform = vtk.vtkTransform()
    transform.Translate(startPoint)  # translate to starting point
    transform.Concatenate(matrix)  # apply direction cosines
    transform.RotateZ(-90.0)  # align cylinder to x axis
    transform.Scale(1.0, length, 1.0)  # scale along the height vector
    transform.Translate(0, .5, 0)  # translate to start of cylinder

    # Transform the polydata
    transformPD = vtk.vtkTransformPolyDataFilter()
    transformPD.SetTransform(transform)
    transformPD.SetInputConnection(cylinderSource.GetOutputPort())

    # Create a mapper and actor for the arrow
    mapper = vtk.vtkPolyDataMapper()
    actor = vtk.vtkActor()
    if USER_MATRIX:
        mapper.SetInputConnection(cylinderSource.GetOutputPort())
        actor.SetUserMatrix(transform.GetMatrix())
    else:
        mapper.SetInputConnection(transformPD.GetOutputPort())
    actor.SetMapper(mapper)
    actor.GetProperty().SetColor(colors.GetColor3d("Cyan"))

    # Create spheres for start and end point
    sphereStartSource = vtk.vtkSphereSource()
    sphereStartSource.SetCenter(startPoint)
    sphereStartSource.SetRadius(0.8)
    sphereStartMapper = vtk.vtkPolyDataMapper()
    sphereStartMapper.SetInputConnection(sphereStartSource.GetOutputPort())
    sphereStart = vtk.vtkActor()
    sphereStart.SetMapper(sphereStartMapper)
    sphereStart.GetProperty().SetColor(colors.GetColor3d("Yellow"))

    sphereEndSource = vtk.vtkSphereSource()
    sphereEndSource.SetCenter(endPoint)
    sphereEndSource.SetRadius(0.8)
    sphereEndMapper = vtk.vtkPolyDataMapper()
    sphereEndMapper.SetInputConnection(sphereEndSource.GetOutputPort())
    sphereEnd = vtk.vtkActor()
    sphereEnd.SetMapper(sphereEndMapper)
    sphereEnd.GetProperty().SetColor(colors.GetColor3d("Magenta"))

    # Create a renderer, render window, and interactor
    renderer = vtk.vtkRenderer()
    renderWindow = vtk.vtkRenderWindow()
    renderWindow.AddRenderer(renderer)
    renderWindow.SetWindowName("Oriented Cylinder")
    renderWindowInteractor = vtk.vtkRenderWindowInteractor()
    renderWindowInteractor.SetRenderWindow(renderWindow)

    # Add the actor to the scene
    renderer.AddActor(actor)
    renderer.AddActor(sphereStart)
    renderer.AddActor(sphereEnd)
    renderer.SetBackground(colors.GetColor3d("BkgColor"))

    # Render and interact
    renderWindow.Render()
    renderWindowInteractor.Start()


if __name__ == '__main__':
    main()

File: src/Java/GeometricObjects/OrientedCylinder.java
import java.awt.BorderLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JPanel;
import javax.swing.SwingUtilities;


import vtk.vtkActor;
import vtk.vtkCylinderSource;
import vtk.vtkNativeLibrary;
import vtk.vtkPanel;
import vtk.vtkPolyDataMapper;
import vtk.vtkMinimalStandardRandomSequence;
import vtk.vtkMath;
import vtk.vtkMatrix4x4;
import vtk.vtkTransform;
import vtk.vtkTransformPolyDataFilter;
import vtk.vtkSphereSource;


//Author: Bharatesh Chakravarthi
//Affiliation: VE Lab, Chung Ang University, Seoul, South Korea. 


public class OrientedCylinder extends JPanel implements ActionListener {
	  private static final long serialVersionUID = 1L;
	  private vtkPanel renWin;
	  private JButton exitButton;

	  // -----------------------------------------------------------------
	  // Load VTK library and print which library was not properly loaded
	  static {
	    if (!vtkNativeLibrary.LoadAllNativeLibraries()) {
	      for (vtkNativeLibrary lib : vtkNativeLibrary.values()) {
	        if (!lib.IsLoaded()) {
	          System.out.println(lib.GetLibraryName() + " not loaded");
	        }
	      }
	    }
	    vtkNativeLibrary.DisableOutputWindow(null);
	  }
	  // -----------------------------------------------------------------
	  
	  
	  public OrientedCylinder() {
	    super(new BorderLayout());

	  // Create a cylinder.
	  // Cylinder height vector is (0,1,0).
	  // Cylinder center is in the middle of the cylinder
	  vtkCylinderSource CylinderSource = new vtkCylinderSource();
	  CylinderSource.SetResolution(15);

	  double sp[] = new double[3];
	  double ep[] = new double[3];
	  
	  // Generate a random start and end point
	  vtkMinimalStandardRandomSequence RandomNumbers = new vtkMinimalStandardRandomSequence();
	  RandomNumbers.SetSeed(8775070);
	  for(int i=0;i<3;++i) {
		  RandomNumbers.Next();
		  sp[i]=RandomNumbers.GetRangeValue(-10, 10);
		  RandomNumbers.Next();
		  ep[i]=RandomNumbers.GetRangeValue(-10, 10);
	  }
	  
	  // Compute a basis
	  double normalizedX[] = new double[3];
	  double normalizedY[] = new double[3];
	  double normalizedZ[] = new double[3];
	  
	  // The X axis is a vector from start to end
	  vtkMath ss = new vtkMath();
	  ss.Subtract(ep, sp, normalizedX);
	  double length;
	  length=ss.Norm(normalizedX);
	  ss.Normalize(normalizedX);
	  
	  // The Z axis is an arbitrary vector cross X
	  double arbitary[] = new double[3];
	  for(int i=0;i<3;++i) {
		  RandomNumbers.Next();
		  arbitary[i]=RandomNumbers.GetRangeValue(-10, 10);
	  }
	  ss.Cross(normalizedX, arbitary, normalizedZ);
	  ss.Normalize(normalizedZ);
	  
	  // The Y axis is Z cross X
	  ss.Cross(normalizedZ, normalizedX, normalizedY);
	  vtkMatrix4x4 Matrix = new vtkMatrix4x4();
	  // Create the direction cosine matrix
	  Matrix.Identity();
	  for(int i=0;i<3;i++) {
		  Matrix.SetElement(i, 0, normalizedX[i]);
		  Matrix.SetElement(i, 1, normalizedY[i]);
		  Matrix.SetElement(i, 2, normalizedZ[i]);
	  }
 
	  vtkTransform Transform = new vtkTransform();
	  Transform.Translate(sp);
	  Transform.Concatenate(Matrix);
	  Transform.RotateZ(-90.0);
	  Transform.Scale(1.0,length,1.0);
	  Transform.Translate(0.0,0.5,0.0);
	  
	  // Apply the transforms
	  vtkTransformPolyDataFilter TransformPD = new vtkTransformPolyDataFilter();
	  TransformPD.SetTransform(Transform);
	  TransformPD.SetInputConnection(CylinderSource.GetOutputPort());
	  
	  // Transform the polydata
	  vtkPolyDataMapper Mapper = new vtkPolyDataMapper();
	  vtkActor Actor = new vtkActor();
	  Mapper.SetInputConnection(CylinderSource.GetOutputPort());


	  Actor.SetMapper(Mapper);
	  Actor.SetUserMatrix(Transform.GetMatrix());
	  Actor.GetProperty().SetColor(1.0,1.0,0.9);
		  
	  vtkSphereSource SphereStartSource = new vtkSphereSource();
	  SphereStartSource.SetCenter(sp);
	  SphereStartSource.SetRadius(0.8);
	  
	  vtkPolyDataMapper SphereStartMapper = new vtkPolyDataMapper();
	  SphereStartMapper.SetInputConnection(SphereStartSource.GetOutputPort());
	  
	  vtkActor SphereStartActor = new vtkActor();
	  SphereStartActor.SetMapper(SphereStartMapper);
	  SphereStartActor.GetProperty().SetColor(1.0,0.0,0.0);
	  
	  vtkSphereSource SphereEndSource = new vtkSphereSource();
	  SphereEndSource.SetCenter(ep);
	  SphereEndSource.SetRadius(0.8);
	  
	  vtkPolyDataMapper SphereEndMapper = new vtkPolyDataMapper();
	  SphereEndMapper.SetInputConnection(SphereEndSource.GetOutputPort());
	  
	  vtkActor SphereEndActor = new vtkActor();
	  SphereEndActor.SetMapper(SphereEndMapper);
	  SphereEndActor.GetProperty().SetColor(0.0,0.0,1.0);

	   
	  renWin = new vtkPanel();
	  renWin.GetRenderer().AddActor(Actor);
	  renWin.GetRenderer().AddActor(SphereStartActor);
	  renWin.GetRenderer().AddActor(SphereEndActor);
	  renWin.GetRenderer().SetBackground(0.105,0.105,0.105);;
	  renWin.resetCamera();
	  

	  // Add Java UI components
	  exitButton = new JButton("Exit");
	  exitButton.addActionListener(this);

	  add(renWin, BorderLayout.CENTER);
	  add(exitButton, BorderLayout.SOUTH);
	  }

	  /** An ActionListener that listens to the button. */
	  public void actionPerformed(ActionEvent e) {
	    if (e.getSource().equals(exitButton)) {
	      System.exit(0);
	    }
	  }

	  public static void main(String s[]) {
	    SwingUtilities.invokeLater(new Runnable() {
	      public void run() {
	        JFrame frame = new JFrame("Oriented Cylinder");
	        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
	        frame.getContentPane().setLayout(new BorderLayout());
	        frame.getContentPane().add(new OrientedCylinder(), BorderLayout.CENTER);
	        frame.setSize(400, 400);
	        frame.setLocationRelativeTo(null);
	        frame.setVisible(true);
	      }
	    });
	  }
	}
